src/backend/gporca/libgpopt/src/base/CUtils.cpp

```diff
@@ -875,6 +875,58 @@ CUtils::FHasCTEAnchor(CExpression *pexpr)
 	return false;
 }
 
+// return CTEConsumers' and a set of CTEProducers' CTE ids in the given subtree
+void
+CUtils::CollectConsumersAndProducers(CMemoryPool *mp, CExpression *pexpr,
+									 ULongPtrArray *cteConsumers, UlongCteIdHashSet *cteProducerSet)
+{
+	COperator *pop = pexpr->Pop();
+
+	if (COperator::EopPhysicalCTEConsumer == pexpr->Pop()->Eopid())
+	{
+		cteConsumers->Append(GPOS_NEW(mp) ULONG(CPhysicalCTEConsumer::PopConvert(pop)->UlCTEId()));
+	}
+	else if (COperator::EopPhysicalCTEProducer == pexpr->Pop()->Eopid())
+	{
+		cteProducerSet->Insert(GPOS_NEW(mp) ULONG(CPhysicalCTEProducer::PopConvert(pop)->UlCTEId()));
+	}
+
+	for (ULONG ul = 0; ul < pexpr->Arity(); ul++)
+	{
+		CExpression *pexprChild = (*pexpr)[ul];
+
+		if (!pexprChild->Pop()->FScalar())
+		{
+			CollectConsumersAndProducers(mp, pexprChild, cteConsumers, cteProducerSet);
+		}
+	}
+}
+
+BOOL
+CUtils::CanReadFromOneSegment(CMemoryPool *mp, CExpression *pexpr)
+{
+       BOOL localityIsValid = true;
```

Comment:
I think we have a mixture of tabs and spaces indentation in this new function.

Reply:
Terribly sorry, fixed once again